### PR TITLE
User Management for GitLab via Artemis

### PR DESCRIPTION
**Pull request by ungaralex**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

**This PR depends on #1110**
For our GitLab+Jenkins setup we don't have any external user management setup like Jira atm. Because of that, we have to manage all users and groups using the Artemis DB.

### Description

- I extended the user management UI allowing admins to also specify the password and edit the groups a user belongs to Those groups now also get displayed in the user component next to the roles. 
- The server now also provides an `ArtemisInternalAuthenticationProvider` in case `jira` and `ldap` are not activated as a Spring profile. In this case, all users get authenticated and managed using the internal Artemis DB. 
- The GitLab service and VCS interface now also allow Artemis to manage and sync the users and groups. This means that any new user gets also added to GitLab and any updated user (e.g. new password) gets also updated in GitLab. Since we also allow instructors to change the groups of a course after creating it, if there are already exercises in this course, the related repositories and groups in GitLab will also get updated (adding/removing new/old instructors/tutors, upgrading/downgrading instructors/tutors). The same goes for changed groups of a user itself. If a user should not be part of an instructor/tutor group after the update, the related entities in GitLab will also be changed.

### Steps for Testing

0. Message me in order to get an account for the testserver
1. Log in to the Artemis Testserver 2
2. Create some new users
3. Those users should also get created in GitLab
4. Create a new course and a programming exercise
5. The users should have the correct rights for the new programming exercise in GitLab (instructors --> Maintainer, tutors --> Guest)
6. Add some more users (tutors, instructors)
7. Those users should now also be a member of the programming exercise group in GitLab
8. Add some users that do not belong to any instructor/tutor group
9. Change the groups of the course, so that some of the new users should now be an instructor/tutor
10. The groups in GitLab should be updated
11. Change the groups of the course, so that some instructors should now only be tutors and check GitLab
12. Change the groups of the course, so that some tutors should now be instructors and check GitLab
13. Change the groups of the course, so that some users are now longer an instructor or tutor and check GitLab

### Screencast
gitla_user_management

### CODE REVIEW

**Comment by krusche** [comment]
Thanks for getting this done. I think there are some changes that can break a lot of behavior, so I think it would be great if you can add some test cases, at least the minimal ones for the functionality that you have changed in this PR so that we really know that nothing important broke.

**Comment by ungaralex** [comment]
I only changed the behavior of the update/create methods in the `UserService` and added some additional logic in case we have no external user management like Jira.

Based on that I added some tests:
- Test the creation of new users
- Test the update of existing users
- Verify, that the (user management portion of the) `VersionControlService` is **not** called for our default configuration (i.e. Bitbucket with Jira). This would be the case for the user create/update/delete operations as well as for the course update endpoint.
- Verify that the new `ArtemisInternalAuthenticationProvider` bean is not available for our default configuration, so it cannot be injected and cause unwanted behavior.

**Comment by ungaralex** [comment]
@krusche If this PR is too big, I can split this up between the client and server changes.

File: package.json [hunk 1]
```diff
@@ -56,7 +56,7 @@
         "ng2-charts": "2.3.0",
         "ng2-dnd": "5.0.2",
         "ngx-cacheable": "1.3.2",
-        "ngx-chips": "2.1.0",
+        "ngx-chips": "^2.1.0",
```
Comment (krusche) [inline package.json]: please revert this change!

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 2]
```diff
@@ -195,7 +202,12 @@ public CourseResource(Environment env, UserService userService, CourseService co
                     return ResponseEntity.badRequest().headers(HeaderUtil.createAlert(applicationName, "The shortname is invalid", "shortnameInvalid")).body(null);
                 }
                 checkIfGroupsExists(updatedCourse);
+                final var oldInstructorGroup = existingCourse.get().getInstructorGroupName();
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java]: please add some comments what this code wants to achieve. It's not really clear to me at the moment why we need this here

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java [hunk 3]
```diff
@@ -66,16 +69,26 @@
...
     private String applicationName;
 
+    @Value("${artemis.external-user-management:true}")
+    private boolean EXTERNAL_USER_MANAGEMENT;
+
     private final UserRepository userRepository;
 
     private final UserService userService;
 
+    private final Optional<ContinuousIntegrationService> continuousIntegrationService;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java]: I am not sure if it is a good idea to have these dependencies in the Resource. I guess we should move all this logic into the UserService

File: package.json (hunk 1 above)
Comment (ungaralex) [reply package.json]: Reverted

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java (hunk 2 above)
Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java]: I added some comments.
Basically, we need to calculate the diff between the old instructors/tutors and the new ones. Based on that, the `VCSUserManagementService` can then remove/add/update users for all exercise repositories in the related course.

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java (hunk 3 above)
Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java]: Yes, this was a bad design choice. I moved the relevant code to the service

File: src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java [hunk 4]
```diff
@@ -97,24 +102,27 @@ public Authentication authenticate(Authentication authentication) throws Authent
...
-        HttpEntity<Principal> entity = new HttpEntity<>(
-                !skipPasswordCheck ? HeaderUtil.createAuthorization(username, password) : HeaderUtil.createAuthorization(JIRA_USER, JIRA_PASSWORD));
-        RestTemplate restTemplate = new RestTemplate();
         ResponseEntity<Map> authenticationResponse = null;
         try {
-            authenticationResponse = restTemplate.exchange(JIRA_URL + "/rest/api/2/user?username=" + username + "&expand=groups", HttpMethod.GET, entity, Map.class);
+            final var path = JIRA_URL + "/rest/api/2/user?username=" + username + "&expand=groups";
+            final HttpEntity<Principal> entity;
+            if (skipPasswordCheck) {
+                authenticationResponse = restTemplate.exchange(path, HttpMethod.GET, null, Map.class);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java]: this code does not have the same logic as before.
`entity  = new HttpEntity<>(HeaderUtil.createAuthorization(JIRA_USER, JIRA_PASSWORD)):` is missing. I guess this is the case, because the REST template already includes this authorization. I wonder however what happens with the REST template if the `else` branch is invoked. Does that change the default authorization?

Also, if the code is correct, you could move `entity` into the else branch completely.

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java]: I guess some comments would be helpful here as well!

File: src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java [hunk 5]
```diff
@@ -215,11 +223,9 @@ public boolean supports(Class<?> authentication) {
     @Override
     public void addUserToGroup(String username, String group) throws ArtemisAuthenticationException {
         log.info("Add user " + username + " to group " + group + " in JIRA");
-        HttpHeaders headers = HeaderUtil.createAuthorization(JIRA_USER, JIRA_PASSWORD);
         Map<String, Object> body = new HashMap<>();
         body.put("name", username);
-        HttpEntity<?> entity = new HttpEntity<>(body, headers);
-        RestTemplate restTemplate = new RestTemplate();
+        HttpEntity<?> entity = new HttpEntity<>(body);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java]: For the new usage of this in JIRA, I think we should call `isGroupAvailable` before to throw a better error.
The super admin can enter any group name in the user management, and the group might not exist in JIRA.

Another option (instead of throwing an error) would be to create the group in such a case.

File: src/main/java/de/tum/in/www1/artemis/service/UserService.java [hunk 6]
```diff
@@ -347,33 +369,52 @@ public void updateUser(String firstName, String lastName, String email, String l
...
+        user = userRepository.save(user);
+        this.clearUserCaches(user);
+
+        forwardUserUpdates(user, oldGroups);
+
+        log.debug("Changed Information for User: {}", user);
+        return user;
+    }
+
+    private void forwardUserUpdates(User user, Set<String> oldGroups) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/UserService.java]: this name sounds strange. What does it mean? please find a more descriptive one
I guess JavaDoc would be good here.

File: src/main/java/de/tum/in/www1/artemis/service/UserService.java [hunk 7]
```diff
@@ -347,33 +369,52 @@ public void updateUser(String firstName, String lastName, String email, String l
...
+        removedGroups.forEach(group -> artemisAuthenticationProvider.removeUserFromGroup(login, group));
+        addedGroups.forEach(group -> artemisAuthenticationProvider.addUserToGroup(login, group));
     }
 
     /**
      * Delete user based on login string
      * @param login user login string
      */
     public void deleteUser(String login) {
+        optionalVcsUserManagementService.ifPresent(userManagementService -> userManagementService.deleteUser(login));
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/UserService.java]: This should not delete a user in the external user management system (e.g. JIRA).

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/UserService.java]: I guess some comments would be good here as well

File: src/test/resources/config/application.yml [hunk 8]
```diff
@@ -2,6 +2,7 @@ artemis:
     repo-clone-path: ./repos/
     repo-download-clone-path: ./repos-download/
     encryption-password: <encrypt-password>     # arbitrary password for encrypting database values
+    external-user-management: true
```
Comment (krusche) [inline src/test/resources/config/application.yml]: this is still outdated, please update the file accordingly

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabUserManagementService.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+            }
+        }
+        catch (GitLabApiException e) {
+            throw new GitLabException("Error while trying to update user in GitLab: " + user, e);
+        }
+    }
+
+    @Override
+    public void updateCoursePermissions(Course updatedCourse, String oldInstructorGroup, String oldTeachingAssistantGroup) {
+        if (oldInstructorGroup.equals(updatedCourse.getInstructorGroupName()) && oldTeachingAssistantGroup.equals(updatedCourse.getTeachingAssistantGroupName())) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabUserManagementService.java]: this method and the one below look quite complex to me. Please add more comments so that it is easier to understand what is happening and more importantly why this is happening!

File: src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java (hunk 4 above)
Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java]: Yes, the authorization gets injected into the `RestTemplate` (Artemis Admin auth). In the else case, we use the user's credentials, because we don't skip the password check. I added some comments to clarify this.

File: src/main/java/de/tum/in/www1/artemis/service/UserService.java (hunk 6 above)
Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/service/UserService.java]: I changed the name and added some JavaDoc

File: src/main/java/de/tum/in/www1/artemis/service/UserService.java (hunk 7 above)
Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/service/UserService.java]: This only deletes the user in the VCS and only if the bean is present (which is the case for GitLab, but not for Bitbucket). I also added some comments

File: src/test/resources/config/application.yml (hunk 8 above)
Comment (ungaralex) [reply src/test/resources/config/application.yml]: Updated

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabUserManagementService.java (hunk 9 above)
Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabUserManagementService.java]: I added lots of new comments. Let me know, what you think

File: src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java [hunk 10]
```diff
@@ -134,14 +145,14 @@ else if (e.getStatusCode().is5xxServerError()) {
...
-            Map content = authenticationResponse.getBody();
-            final String login = (String) content.get("name");
-            final String emailAddress = (String) content.get("emailAddress");
-            User user = userRepository.findOneByLogin(login).orElseGet(() -> userService.createUser(login, "", (String) content.get("displayName"), "", emailAddress, null, "en"));
-            user.setEmail(emailAddress);
+        if (authenticationResponse != null && authenticationResponse.getBody() != null) {
+            final var content = authenticationResponse.getBody();
+            final var login = content.getName();
+            final var emailAddress = content.getEmailAddress();
+            final var user = userRepository.findOneByLogin(login).orElseGet(() -> userService.createUser(login, password, content.getDisplayName(), "", emailAddress, null, "en"));
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java]:
```suggestion
            final var user = userRepository.findOneByLogin(login).orElseGet(() -> userService.createUser(login, "", content.getDisplayName(), "", emailAddress, null, "en"));
```

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java]: we don't want to store the password of JIRA users in our db!!!

Comment (ungaralex) [reply src/main/java/de/tum/in/www1/artemis/security/JiraAuthenticationProvider.java]: reverted
